programs/bpf_loader/src/lib.rs

```diff
@@ -620,37 +632,70 @@ fn process_loader_upgradeable_instruction(
-                    return Err(InstructionError::MissingRequiredSignature);
+            }
+
+            match close_account.state()? {
+                UpgradeableLoaderState::Uninitialized => {
+                    recipient_account
+                        .try_account_ref_mut()?
+                        .checked_add_lamports(close_account.lamports()?)?;
+                    close_account.try_account_ref_mut()?.set_lamports(0);
                 }
+                UpgradeableLoaderState::Buffer { authority_address }
+                | UpgradeableLoaderState::ProgramData {
+                    slot: _,
+                    upgrade_authority_address: authority_address,
+                } => {
+                    let authority = keyed_account_at_index(keyed_accounts, 2)?;
+
+                    if authority_address.is_none() {
+                        ic_logger_msg!(logger, "Account is immutable");
+                        return Err(InstructionError::Immutable);
+                    }
+                    if authority_address != Some(*authority.unsigned_key()) {
+                        ic_logger_msg!(logger, "Incorrect authority provided");
+                        return Err(InstructionError::IncorrectAuthority);
+                    }
+                    if authority.signer_key().is_none() {
+                        ic_logger_msg!(logger, "Authority did not sign");
+                        return Err(InstructionError::MissingRequiredSignature);
+                    }
 
-                recipient_account
-                    .try_account_ref_mut()?
-                    .checked_add_lamports(close_account.lamports()?)?;
-                close_account.try_account_ref_mut()?.set_lamports(0);
-                for elt in close_account.try_account_ref_mut()?.data_as_mut_slice() {
-                    *elt = 0;
+                    recipient_account
+                        .try_account_ref_mut()?
+                        .checked_add_lamports(close_account.lamports()?)?;
+                    close_account.try_account_ref_mut()?.set_lamports(0);
+                    if !invoke_context.is_feature_active(&close_upgradeable_program_accounts::id())
+                    {
+                        for elt in close_account.try_account_ref_mut()?.data_as_mut_slice() {
+                            *elt = 0;
+                        }
+                    } else {
+                        close_account.set_state(&UpgradeableLoaderState::Uninitialized)?;
+                    }
+                }
+                _ => {
+                    ic_logger_msg!(logger, "Account does not support closing");
+                    return Err(InstructionError::InvalidArgument);
                 }
-            } else {
-                ic_logger_msg!(logger, "Account does not support closing");
-                return Err(InstructionError::InvalidArgument);
             }
 
             ic_logger_msg!(logger, "Closed {}", close_account.unsigned_key());
```

Comment:
What do you think of changing this key to the program id if the programdata is being closed? Also, it would be nice if the log specified what type of account was closed.

Reply:
We don't know the program id at this point, there is no reverse lookup and I wasn't planning on changing the close instruction to include the extra Program account.

Reply:
After making the changes to require the program write lock I also made these changes.